logging: Add rate limiting support

krish2718 opened the pull request:
Add ratelimiited variants for use in data paths.

krish2718 committed:
logging: Add ratelimited variants

These are helpful for logging in data-path, not to overwhelm the
console but still keep the user informed about the issues.

This is similar to printk_ratelimited in Linux kernel.

Add only basic build tests for now.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

krish2718 committed:
net: ip: Use rate limited version

As this is the hot data path, use a rate limited warning variant.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

krish2718 committed:
logging: Fix thread safety for warn once

Use atomic variables to make it thread safe.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

pdgendt reviewed (commented):
Very useful feature!

JordanYates commented on include/zephyr/logging/log.h:
```diff
@@ -89,13 +90,189 @@ extern "C" {
+ */
+#define LOG_WRN_RATELIMITED_RATE(_rate_ms, ...)					\
+	do {									\
+		static atomic_t __last_log_time = ATOMIC_INIT(0);		\
```
statics shouldn't be initialised to `0`, that happens automatically.
Same for all the other macros.

JordanYates commented on include/zephyr/logging/log.h:
```diff
@@ -89,13 +90,189 @@ extern "C" {
+		uint32_t __last = atomic_get(&__last_log_time);		\
+		if (unlikely(__now - __last >= (_rate_ms))) {		\
+			if (atomic_cas(&__last_log_time, __last, __now)) {	\
+				Z_LOG(LOG_LEVEL_WRN, __VA_ARGS__);		\
```
Not sure if this is possible or even desired, but in my head it would be great if the message output the number of skipped prints since the previous output.
```
Z_LOG(LOG_LEVEL_WRN, "Skipped %d: " GET_ARG_N(0, __VA_ARGS__), __skipped, GET_ARGS_LESS_N(1, __VA_ARGS__));
```

pdgendt commented on doc/releases/release-notes-4.3.rst:
```diff
@@ -65,6 +65,23 @@ New APIs and options
 
 .. zephyr-keep-sorted-start re(^\* \w)
 
+* :c:macro:`LOG_ERR_RATELIMITED` - Rate-limited error logging macro (convenience)
+* :c:macro:`LOG_WRN_RATELIMITED` - Rate-limited warning logging macro (convenience)
+* :c:macro:`LOG_INF_RATELIMITED` - Rate-limited info logging macro (convenience)
+* :c:macro:`LOG_DBG_RATELIMITED` - Rate-limited debug logging macro (convenience)
+* :c:macro:`LOG_HEXDUMP_ERR_RATELIMITED` - Rate-limited error hexdump macro (convenience)
+* :c:macro:`LOG_HEXDUMP_WRN_RATELIMITED` - Rate-limited warning hexdump macro (convenience)
+* :c:macro:`LOG_HEXDUMP_INF_RATELIMITED` - Rate-limited info hexdump macro (convenience)
+* :c:macro:`LOG_HEXDUMP_DBG_RATELIMITED` - Rate-limited debug hexdump macro (convenience)
+* :c:macro:`LOG_ERR_RATELIMITED_RATE` - Rate-limited error logging macro (explicit rate)
+* :c:macro:`LOG_WRN_RATELIMITED_RATE` - Rate-limited warning logging macro (explicit rate)
+* :c:macro:`LOG_INF_RATELIMITED_RATE` - Rate-limited info logging macro (explicit rate)
+* :c:macro:`LOG_DBG_RATELIMITED_RATE` - Rate-limited debug logging macro (explicit rate)
+* :c:macro:`LOG_HEXDUMP_ERR_RATELIMITED_RATE` - Rate-limited error hexdump macro (explicit rate)
+* :c:macro:`LOG_HEXDUMP_WRN_RATELIMITED_RATE` - Rate-limited warning hexdump macro (explicit rate)
+* :c:macro:`LOG_HEXDUMP_INF_RATELIMITED_RATE` - Rate-limited info hexdump macro (explicit rate)
+* :c:macro:`LOG_HEXDUMP_DBG_RATELIMITED_RATE` - Rate-limited debug hexdump macro (explicit rate)
```
`RATELIMITED` still needs to change to `RATELIMIT` here.

pdgendt commented on include/zephyr/logging/log.h:
```diff
@@ -89,13 +90,318 @@ extern "C" {
+#else /* CONFIG_LOG_RATELIMIT */
+
+/* Rate-limited logging macros when CONFIG_LOG_RATELIMIT is disabled */
+#define LOG_WRN_RATELIMIT_RATE(_rate_ms, ...) (void) 0
+#define LOG_WRN_RATELIMIT(...) (void) 0
+#define LOG_ERR_RATELIMIT_RATE(_rate_ms, ...) (void) 0
+#define LOG_ERR_RATELIMIT(...) (void) 0
+#define LOG_INF_RATELIMIT_RATE(_rate_ms, ...) (void) 0
+#define LOG_INF_RATELIMIT(...) (void) 0
+#define LOG_DBG_RATELIMIT_RATE(_rate_ms, ...) (void) 0
+#define LOG_DBG_RATELIMIT(...) (void) 0
+#define LOG_HEXDUMP_ERR_RATELIMIT_RATE(_rate_ms, _data, _length, _str) (void) 0
+#define LOG_HEXDUMP_ERR_RATELIMIT(_data, _length, _str) (void) 0
+#define LOG_HEXDUMP_WRN_RATELIMIT_RATE(_rate_ms, _data, _length, _str) (void) 0
+#define LOG_HEXDUMP_WRN_RATELIMIT(_data, _length, _str) (void) 0
+#define LOG_HEXDUMP_INF_RATELIMIT_RATE(_rate_ms, _data, _length, _str) (void) 0
+#define LOG_HEXDUMP_INF_RATELIMIT(_data, _length, _str) (void) 0
+#define LOG_HEXDUMP_DBG_RATELIMIT_RATE(_rate_ms, _data, _length, _str) (void) 0
+#define LOG_HEXDUMP_DBG_RATELIMIT(_data, _length, _str) (void) 0
```
Maybe overkill but could have a fallback option:

```
choice LOG_RATELIMIT_FALLBACK
	prompt "Rate-limited logging fallback"
	default LOG_RATELIMIT_DROP

config LOG_RATELIMIT_DROP
	bool "Drop ..."

config LOG_RATELIMIT_ONCE
	bool "Log once"

config LOG_RATELIMIT_FULL
	bool "Log all ..."

endchoice
```

Or... don't make it optional and make it part of the logging.

krish2718 replied:
I did something like this:

```
if !LOG_RATELIMIT
choice f
	prompt "Fallback behavior for rate-limited logging macros"
	default LOG_RATELIMIT_FALLBACK_DROP
	help
	  Select fallback behavior for rate-limited logging macros when
	  LOG_RATELIMIT is disabled.

	  - "Log all prints": All log prints are output, i.e., rate limiting is ignored.
	  - "Drop all prints": All log prints using rate-limited macros are dropped.

	  This option allows you to control whether rate-limited log macros
	  should always print or be completely suppressed when rate limiting
	  is not available.

config LOG_RATELIMIT_FALLBACK_LOG
	bool "Log all prints (ignore rate limiting)"
	help
	  All log prints using rate-limited macros are output as normal logs.

config LOG_RATELIMIT_FALLBACK_DROP
	bool "Drop all prints"
	help
	  All log prints using rate-limited macros are dropped (no output).

endchoice
```

JordanYates commented on include/zephyr/logging/log.h:
```diff
@@ -89,13 +90,351 @@ extern "C" {
+		uint32_t __now = k_uptime_get_32();				\
+		uint32_t __last = atomic_get(&__last_log_time);		\
+		uint32_t __diff = __now - __last;				\
+		if (unlikely(__diff >= (_rate_ms))) {			\
```
Since this is expected to be called under heavy loops, we can avoid the diff calculation on each loop by precomputing the next time.
```
static  ___next;
uint32_t __now = k_uptime_get_32();
if (unlikely(now > __next)) {
   __next = now + _rate_ms;
    ...
}
```

JordanYates commented on include/zephyr/logging/log.h:
```diff
@@ -89,13 +90,351 @@ extern "C" {
+		uint32_t __diff = __now - __last;				\
+		if (unlikely(__diff >= (_rate_ms))) {			\
+			if (atomic_cas(&__last_log_time, __last, __now)) {	\
+				uint32_t __skipped = atomic_get(&__skipped_count);	\
```
`atomic_clear` returns the previous value of the variable.
```
uint32_t __skipped = atomic_clear(&__skipped_count);
```

pdgendt commented on include/zephyr/logging/log.h:
```diff
@@ -89,13 +90,351 @@ extern "C" {
+#define LOG_HEXDUMP_DBG_RATELIMIT(_data, _length, _str) \
+	LOG_HEXDUMP_DBG_RATELIMIT_RATE(CONFIG_LOG_RATELIMIT_INTERVAL_MS, _data, _length, _str)
+
+#else /* CONFIG_LOG_RATELIMIT */
+
+/* Rate-limited logging macros when CONFIG_LOG_RATELIMIT is disabled */
+#if defined(CONFIG_LOG_RATELIMIT_FALLBACK_LOG)
```
Nit: can be combined in an `#elif`

krish2718 commented on include/zephyr/logging/log.h:
```diff
@@ -89,13 +90,350 @@ extern "C" {
+ */
+#define _LOG_RATELIMIT_CORE(_level, _rate_ms, _skipped_msg, _log_func, ...)	\
+	do {									\
+		static atomic_t __last_log_time ;		\
+		static atomic_t __skipped_count ;		\
+		uint32_t __now = k_uptime_get_32();				\
+		uint32_t __last = atomic_get(&__last_log_time);		\
+		uint32_t __next = __last + (_rate_ms);			\
+		if (unlikely(__now >= __next)) {				\
```
Urgh, missed that, good catch, will revert to the diff approach.

krish2718 replied:
Yeah, I am fine with switching to seconds resolution, I don't think we need the `ms` granularity, went with this because it caters to both ms or s usecases.

krish2718 committed:
net: Add ratelimited variants

Add the ratelimtied variants for data path debug.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

krish2718 commented:
1. Fixed changes that went to wrong commit
2. Fixed a case where logging wasn't enabled, and as we are using below logic the macros won't be defined, so, rejig to ensure they are always defined, this fixes a build error (caught in downstream)

```
#ifdef _LOG
#elif _DROP
#end
```

nordic-krch commented on include/zephyr/logging/log.h:
```diff
@@ -89,13 +90,346 @@ extern "C" {
+ */
+#define _LOG_RATELIMIT_CORE(_level, _rate_ms, _skipped_msg, _log_func, ...)	\
+	do {									\
+		static atomic_t __last_log_time ;		\
```
can you add support for CONFIG_LOG_RATELIMIT_FALLBACK_DROP and  CONFIG_LOG_RATELIMIT inside this macro as `if(IS_ENABLED( CONFIG_LOG_RATELIMIT) ...` instead of using ifdefs. That is the approach in all logging macros. It ensures that there is no warning about unused variables when logging is disabled and variables are only used in logging macros.

nordic-krch replied:
No, `COND_CODE_1(a, (b), (c))` is resolved by preprocessor and only `b` or `c` is seen by the compiler. When `if (IS_ENABLED(a) b else c` is used then both `b` and `c` are compiled (but one is cut).

nordic-krch reviewed (commented):
Looks ok, one minor comment.

pdgendt commented on include/zephyr/logging/log.h:
```diff
@@ -89,13 +91,334 @@ extern "C" {
+ * @param _log_func Logging function to call (Z_LOG or Z_LOG_HEXDUMP)
+ * @param ... Arguments to pass to the logging function.
+ */
+#define _LOG_RATELIMIT_CORE(_level, _rate_ms, _skipped_msg, _log_func, ...)	\
```
Do you need to pass the `_skipped_msg`? It's the same now for all calling macros.

pdgendt commented on include/zephyr/logging/log.h:
```diff
@@ -89,13 +91,455 @@ extern "C" {
+ * @param ... A string optionally containing printk valid conversion specifier,
+ * followed by as many values as specifiers.
+ */
+#define LOG_WRN_RATELIMIT(...) \
+	COND_CODE_1(CONFIG_LOG_RATELIMIT, \
+		(_LOG_RATELIMIT_LVL(LOG_LEVEL_WRN, CONFIG_LOG_RATELIMIT_INTERVAL_MS, \
+				   __VA_ARGS__)), \
+		(COND_CODE_1(CONFIG_LOG_RATELIMIT_FALLBACK_DROP, \
+			((void)0), \
+			(LOG_WRN(__VA_ARGS__)))))
```
I fail to understand the need for `COND_CODE_1` here? Is it because of an undefined `CONFIG_LOG_RATELIMIT_INTERVAL_MS`?

How about:
```c
#ifdef CONFIG_LOG_RATELIMIT
#define LOG_RATELIMIT_INTERVAL_MS CONFIG_LOG_RATELIMIT_INTERVAL_MS
#else
#define LOG_RATELIMIT_INTERVAL_MS 0
#endif

// ...
#define LOG_WRN_RATELIMIT(...) \
	_LOG_RATELIMIT_LVL(LOG_LEVEL_WRN, LOG_RATELIMIT_INTERVAL_MS, __VA_ARGS__)

krish2718 replied:
>  Is it because of an undefined CONFIG_LOG_RATELIMIT_INTERVAL_MS?

yes.

> How about:

Yeah, we can try this as well,

pdgendt reviewed (dismissed):
Very nice! Just some nits.

krish2718 commented on include/zephyr/logging/log.h:
```diff
@@ -89,13 +97,387 @@ extern "C" {
+		} else {							\
+			atomic_inc(&__skipped_count);				\
+		}								\
+	} while (0)
```
Yes, to fix long lines had to take a few shortcuts :)

krish2718 replied:
Looking the rest, I don't see a fixed style or patter, in the same file, tabs + spaces even in a single line are used and they don't necessarily all align.

krish2718 replied:
IMHO, best thing to do here is `clang-format -i include/zephyr/logging/log.h` and create a new commit? WDYT?

pdgendt replied:
I'm not convinced it's the best to do here 😅, it would add more churn into the PR. Keeping it consistent for the PR should be sufficient.

nordic-krch commented on include/zephyr/logging/log.h:
```diff
@@ -87,15 +97,461 @@ extern "C" {
+	do {                                                                                       \
+		static atomic_t __last_log_time;                                                   \
+		static atomic_t __skipped_count;                                                   \
+		uint32_t __now = k_uptime_get_32();                                                \
```
that would require `#include <zephyr/kernel.h>` but I would prefer that it's not included in `log.h` as we try to include as little as possible to allow including `log.h` anywhere. Maybe extern?

nordic-krch replied:
maybe a little hacky `extern uint32_t arch_k_cycle_get_32(void)`?

krish2718 replied:
Even that is static line, but I see `sys_clock_cycle_get_32` can be used, WDYT?

krish2718 replied:
But then we'd need `k_ticks_to_ms_floor64` which needs another header.

krish2718 committed:
logging: Fix formatting

Auto-format using clang-format, esp. for the aligning backslahes and a
couple of manual edits to fix checkpatch.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>
